pkg/skaffold/runner/runner.go

```diff
@@ -205,67 +204,68 @@ func getTagger(t latest.TagPolicy, customTag string) (tag.Tagger, error) {
 	}
 }
 
+func (r *SkaffoldRunner) newLogger(out io.Writer, artifacts []*latest.Artifact) *kubernetes.LogAggregator {
+	return kubernetes.NewLogAggregator(out, imageNames(artifacts), r.imageList)
+}
+
 // Run builds artifacts, runs tests on built artifacts, and then deploys them.
 func (r *SkaffoldRunner) Run(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) error {
-	bRes, err := r.Build(ctx, out, r.Tagger, artifacts)
-	if err != nil {
-		return errors.Wrap(err, "build step")
+	if err := r.buildTestDeploy(ctx, out, artifacts); err != nil {
+		return err
 	}
 
-	if err = r.Test(ctx, out, bRes); err != nil {
-		return errors.Wrap(err, "test step")
-	}
+	if r.opts.Tail {
+		logger := r.newLogger(out, artifacts)
+		if err := logger.Start(ctx); err != nil {
+			return errors.Wrap(err, "starting logger")
+		}
 
-	_, err = r.Deploy(ctx, out, bRes)
-	if err != nil {
-		return errors.Wrap(err, "deploy step")
+		<-ctx.Done()
 	}
 
-	return r.TailLogs(ctx, out, artifacts, bRes)
+	return nil
 }
 
-// TailLogs prints the logs for deployed artifacts.
-func (r *SkaffoldRunner) TailLogs(ctx context.Context, out io.Writer, artifacts []*latest.Artifact, bRes []build.Artifact) error {
-	if !r.opts.Tail {
-		return nil
+func (r *SkaffoldRunner) buildTestDeploy(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) error {
+	bRes, err := r.Build(ctx, out, r.Tagger, artifacts)
+	if err != nil {
+		return errors.Wrap(err, "build failed")
 	}
 
-	imageList := kubernetes.NewImageList()
-	for _, b := range bRes {
-		imageList.Add(b.Tag)
+	if err = r.Test(ctx, out, bRes); err != nil {
```

Comment:
How about skip if test is nil and get rid of noop tester?

Reply:
And the helper function

Reply:
`if tests is nil`? Do you mean we'd have to store the `SkaffoldPipeline` in the SkaffoldRunner instance?